### DISCUSSION TITLE
Is there a way to use the webworker API in a Web extension?

### DISCUSSION THREAD

**Discussion by fbricon** [microsoft/vscode-discussions#420] [Extension Development QnA] [answered]
Is it possible to access the webworker APIs when running as a web extension? I'd like to be able to collect some statistics from [WorkerNavigator](https://developer.mozilla.org/en-US/docs/Web/API/WorkerNavigator), and also be able to use the [indexedDB API](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API).

When I try to add @types/web to my project that already depends on @types/node, the compiler complains:

```
node_modules/@types/web/index.d.ts:2074:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; }', but here has type '{ new (): AbortSignal; prototype: AbortSignal; timeout(milliseconds: number): AbortSignal; }'.

2074 declare var AbortSignal: {
                 ~~~~~~~~~~~

  node_modules/@types/node/ts4.8/globals.d.ts:114:13
    114 declare var AbortSignal: {
                    ~~~~~~~~~~~
    'AbortSignal' was also declared here.


Found 1 error in node_modules/@types/web/index.d.ts:2074
```

**Comment by fbricon** [comment]
VS Code has its own internal [IndexedDB API](https://github.com/microsoft/vscode/blob/920540cba4270dfe35817ae53c50ce921cbda706/src/vs/base/browser/indexedDB.ts#L23), but I don't think it's accessible to extensions.

**Comment by jrieken** [comment, marked as answer]
Yeah, that should be possible. The WebWorker extension host only disables a few native APIs but things like indexeddb or creating another worker should be possible. This is how I do it in [anycode](https://insiders.vscode.dev/github/microsoft/vscode-anycode/blob/d70c9d09eacc511b151ef98e874a29d98a332090/anycode/server/src/browser/storage.ts#L24) - it stores an index of symbols per workspace. 

Tho, having said that there might be of dts-challenges. Don't let them scare you and remember that webworker != web

**Comment by fbricon** [comment]
@jrieken thanks for this useful pointer. Now I see that I need to separate my code into different modules with different tsconfigs.

PS: I had some issues with your insiders.vscode.dev link. Direct link to the github repo works better : https://github.com/microsoft/vscode-anycode/blob/d70c9d09eacc511b151ef98e874a29d98a332090/anycode/server/src/browser/storage.ts#L24

**Comment by jrieken** [comment]
Glad that I could help but also curious how the vscode-dev link failed? Was there failure or the need to login that turned you away from it?

**Comment by fbricon** [comment]
Page never loaded on 3 browsers. It eventually worked several minutes later when I retried.
